src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs

```diff
@@ -235,9 +235,9 @@ private void SendKmsRequest(KmsRequest request, CancellationToken cancellation)
                 var requestBytes = request.Message.ToArray();
                 sslStream.Write(requestBytes);
 
-                var buffer = new byte[4096];
                 while (request.BytesNeeded > 0)
                 {
+                    var buffer = new byte[request.BytesNeeded]; // BytesNeeded is the maximum number of bytes that libmongocrypt wants to receive.
```

Comment:
Fixing an old bug. This will cause a problem each time then a kms response is bigger than 1024 bytes. Previously we didn't have such big responses, but now, the azure returns more than 2000 bytes in response.

Reply:
I'm not sure I understand.  The old buffer was 4096 bytes, so why would 2000 bytes cause a problem?  Also, wouldn't the existing code just continue to feed the response bytes until it was done?

Reply:
it's related to behavior in `mongocrypt.dll`. We discussed this with Kevin https://mongodb.slack.com/archives/C016YF807JS/p1602889908019400 and offline. The buffer size which we pass into `libmongocrypt` is important. And in case if we pass more than 2000 bytes (and more than libmongocrypt expects), it affects inner library logic, I don't remember what was exactly wrong in `c` library, we can discuss it with him

Reply:
The old buffer had a capacity of 4096 bytes, but was not being completely filled when only AWS responses were being returned. Azure returns larger responses. libmongocrypt explicitly errors if it is fed more than the number of bytes it requests, which is documented in the API: https://github.com/mongodb/libmongocrypt/blob/7a53c00f8c80a672c5c6b417d4418da0cccb2acb/src/mongocrypt.h.in#L908:L909.

The initial amount it requests is 1024 bytes (which is arbitrary). And in discussing with @DmitryLukyanov, I think we can remove the arbitrary limit and error in libmongocrypt. I filed https://jira.mongodb.org/browse/MONGOCRYPT-286 to do so. But it seemed more safer to me to have C# follow the current prescribed API now rather than change libmongocrypt's behavior.